pkg/openthread: add FTD-MTD-NCP tests

biboc opened the pull request:
This ~WIP~ PR implements OpenThread Full Thread Device (FTD), Minimal Thread Device (MTD) and Network Co-Processor (NCP) in RIOT.

FTD is a Thread device with routing-capability.
MTD does not have routing capability.
NCP is able to communicate with wpantund on a linux machine by UART.

biboc commented:
It's a WIP PR because NCP does not work anymore and I've discovered a bug in our port of OpenThread.

The problem is still linked with otTasklets. 
In OpenThread example, there is only one context with a while loop:
```
   while (1)
    {
        otTaskletsProcess(sInstance);
        PlatformProcessDrivers(sInstance);
    }
```
`PlatformProcessDrivers(sInstance);` checks if there is a radio packet, uart packet to send/receive and alarm that fired. So in contrary to RIOT, it pulls as fast as possible these three peripheral and at the same time execute pending task with `otTaskletsProcess(sInstance);`

By putting `otTaskletsProcess(sInstance)` in `otTaskletsSignalPending()` function, we might encounter an recusirve call because `otTaskletsProcess()` might call `otTaskletsSignalPending()` !!

So I don't really know where to add `otTaskletsProcess(sInstance)` since RIOT is event driven and does not have crazy while loop.

Is it clear?
@aabadie @jia200x

jia200x commented:
@biboc Great to see this PR! I'm working in the sock part.

I see the problem. I just saw [this](https://groups.google.com/forum/#!topic/openthread-users/NNXg3CSZpvk). Maybe a msg_send_self in otTaskletsSignalPending could resolve this situation?

I will investigate further.

jia200x commented:
@biboc could you try something like:

```
bool execute_taks=false;
void otTaskletsProcess(sInstance)
{
    execute_task = true;
}

...
msg_receive(&msg);
if(execute_task)
{
    otTaskletsProcess(sInstance);
    execute_task = false;
}
switch(...)
...
```

When otTasklestsProcess was in the beginning of the while loop, it was executed each time a msg arrived. I'm not sure if adding a flag resolves the issue, but I would give it a try.

Cheers

biboc commented:
Something like this should do like POSIX example does:

```
while (1) {
        otTaskletsProcess(sInstance);
         if (!otTaskletsArePending(sInstance)) {
            msg_receive(&msg);
            switch (msg.type) {
                case OPENTHREAD_XTIMER_MSG_TYPE_EVENT:
                   ....
                case OPENTHREAD_JOB_MSG_TYPE_EVENT:
                    job = msg.content.ptr;
                    reply.content.value = ot_exec_command(sInstance, job->command, job->arg, job->answer);
                    msg_reply(&msg, &reply);
                    break;
            }
         }
    }
```

But still it does not work as expected

biboc commented:
My last edits work great! This is the right way to implement otTask.
Now, NCP is not working as expected. Wpantund detects the NCP, it can send commands but "scan" and "form" commands fail. Can you have a try @jia200x?

biboc commented:
Works better now! 
Scan is working, form a network is still broken

jia200x commented:
Hi @biboc.

For some reason I'm have 2 leader nodes with the openthread_ftd example. I'm investigating what might cause that.

I will do some NCP tests as well.

jia200x commented:
@biboc the scan is working. But, as said before, there's something wrong with the core. That could be related to the "form" problem.

Cheers

jia200x commented:
@biboc after some attempts, 2 ftd nodes are working ok. I will test it on Iot-LAB

Cheers

jia200x commented:
@biboc just discovered something strange:

There are some commands I'm unable to execute with the OpenThread CLI.
E.g:

```
extpanid DEBA7AB1E5EAF00D
Error 14: InvalidState
```

Have you had this behavior before?

biboc commented:
To change the panid you need to "thread stop" before

biboc commented:
Your test is great help, thanks

biboc commented:
I upgraded OpenThread to the last version.
@jia200x, can you have a look at platform_radio.c: send_pkt(): otPlatRadioTxDone(aInstance, &sTransmitFrame, NULL, OT_ERROR_NONE);
I use NULL since I can't get the ACK frame from netdev (even with at86rf2xx since it is automated) 
What do you think?

jia200x commented:
hi @biboc 

If I'm not wrong, the NETDEV_EVENT_TX_COMPLETE should be triggered only if there's an ACK (@miri64 @OlegHahm ??). Otherwise, the NETDEV2_EVENT_NO_ACK will be triggered.
What's the behavior you are having with ACK?

biboc commented:
The problem here is the change in OpenThread API, instead of using `otPlatRadioTransmitDone`, we should use `otPlatRadioTxDone`:

```
/**
 * The radio driver calls this method to notify OpenThread that the transmission has completed,
 * this callback pass up the ACK frame, new add platforms should use this callback function.
 *
 * @param[in]  aInstance  The OpenThread instance structure.
 * @param[in]  aFrame     A pointer to the frame that was transmitted.
 * @param[in]  aAckFrame  A pointer to the ACK frame, NULL if no ACK was received.
 * @param[in]  aError     OT_ERROR_NONE when the frame was transmitted, OT_ERROR_NO_ACK when the frame was
 *                        transmitted but no ACK was received, OT_ERROR_CHANNEL_ACCESS_FAILURE when the transmission
 *                        could not take place due to activity on the channel, OT_ERROR_ABORT when transmission was
 *                        aborted for other reasons.
 *
 */
extern void otPlatRadioTxDone(otInstance *aInstance, otRadioFrame *aFrame, otRadioFrame *aAckFrame,
                              otError aError);

/**
 * The radio driver calls this method to notify OpenThread that the transmission has completed,
 * this function is going to be deprecated, new add platfroms should not use this callback function.
 *
 * @param[in]  aInstance      The OpenThread instance structure.
 * @param[in]  aFrame         A pointer to the frame that was transmitted.
 * @param[in]  aFramePending  TRUE if an ACK frame was received and the Frame Pending bit was set.
 * @param[in]  aError         OT_ERROR_NONE when the frame was transmitted, OT_ERROR_NO_ACK when the frame was
 *                            transmitted but no ACK was received, OT_ERROR_CHANNEL_ACCESS_FAILURE when the transmission
 *                            could not take place due to activity on the channel, OT_ERROR_ABORT when transmission was
 *                            aborted for other reasons.
 *
 */
extern void otPlatRadioTransmitDone(otInstance *aInstance, otRadioFrame *aFrame, bool aFramePending,
                                    otError aError);
```

So I need to pass the ACK frame as mentioned here: 
` * @param[in]  aAckFrame  A pointer to the ACK frame, NULL if no ACK was received.` 
But netdev don't provide it.

Is it clearer? @jia200x

biboc commented:
We've fixed UART problem for NCP with buffering the current frame before giving it to OpenThread, it works great!

Now I would like @jia200x to have a look at the `otPlatRadioTxDone` in my previous comment so we can end this PR

jia200x commented:
hi @biboc.
Sorry for the delay, I was busy with work.

Your buffering solution is very smart! Congratz.

Concerning the ACK part, IMHO I would enable OPENTHREAD_CONFIG_LEGACY_TRANSMIT_DONE and use the otPlatRadioTransmitDone, until we find another solution. There's no way to retrieve the ACK from the driver with Auto ACK. Also, passing Ack=NULL [may have side effects](https://github.com/openthread/openthread/blob/9a3798471559ec5ee6cb335e3a179ea9f8f0a245/src/core/mac/mac.cpp#L983).

Cheers!

biboc commented:
Yes they use it only for `ackFrame->GetPower()` finally but you're right let's use old function as long as we can

gebart commented:
So it's not the ACK itself, but the LQI or RSSI of the ACK that you need?

biboc commented:
@gebart, it might be only for this yes but since we can still use old function I'll vote to use old function and not "hack" the function by filling only receiving power and even this I'm not sure power is updated when an ACK is received.

biboc commented:
@aabadie 
Can you review this PR? 
I'll squash commits after your review

aabadie commented:
> Can you review this PR?

I didn't test but do you confirm that it works now ?

aabadie commented on cpu/sam0_common/periph/i2c.c:
```diff
@@ -11,7 +11,7 @@
  * @file
  * @brief       Low-level I2C driver implementation
- * @author      Baptiste Clenet <baptiste.clenet@xsoen.com>
+ * @author      Baptiste Clenet <bapclenet@gmail.com>
```
I think this change could be part of a separated PR (that will be fast to review)

biboc replied:
Yes I was lazy to do two PRs...

aabadie replied:
The problem is that it touches very different files (in sam0_common, samd21 and saml21). Can you open a dedicated PR for these changes ?

biboc replied:
ok done

aabadie commented on cpu/samd21/periph/rtc.c:
```diff
@@ -11,7 +11,7 @@
  * @file
  * @brief       Low-level RTC driver implementation
- * @author      Baptiste Clenet <baptiste.clenet@xsoen.com>
+ * @author      Baptiste Clenet <bapclenet@gmail.com>
```
same here

aabadie commented on cpu/saml21/periph/rtc.c:
```diff
@@ -11,7 +11,7 @@
  * @file
  * @brief       Low-level RTC driver implementation
- * @author      Baptiste Clenet <baptiste.clenet@xsoen.com>
+ * @author      Baptiste Clenet <bapclenet@gmail.com>
```
and here

aabadie commented on pkg/openthread/Makefile:
```diff
@@ -1,11 +1,22 @@
+endif
+ifneq (,$(filter libopenthread-cli-mtd,$(USEMODULE)))
+    $(info Compile OpenThread for MTD device)
+    OPENTHREAD_ARGS += --enable-cli-app=mtd  --enable-joiner
```
extra space before `--enable-joiner`

aabadie commented on pkg/openthread/Makefile:
```diff
@@ -1,11 +1,22 @@
+    $(info Compile OpenThread with NCP)
+    OPENTHREAD_ARGS += --enable-ncp-app=ftd --with-ncp-bus=uart --enable-commissioner
+endif
+OPENTHREAD_ARGS     += --enable-application-coap
```
no need to align here

aabadie commented on tests/openthread_ncp/main.c:
```diff
@@ -0,0 +1,23 @@
+
+int main(void)
+{
+    return 0;
```
This looks weird, I though one could use the shell or have some welcome message here ?

biboc replied:
The shell is automatically run here, I'll add a welcome message

aabadie commented on tests/openthread_ncp/README.md:
```diff
@@ -0,0 +1,43 @@
+3. Start wpantund on your host.  Be sure to pass flags appropriate to connect wpantund to your NCP connection type.
+You can also name the network interface at this time or let wpantund automatically assign a name (wpan0):
+
+`sudo /usr/local/sbin/wpantund -s /dev/ttyACM0 -I wpan0 -o SyslogMask all`
```
use triple back quotes

aabadie commented on tests/openthread_ncp/README.md:
```diff
@@ -0,0 +1,43 @@
+
+To test OpenThread NCP in RIOT, you can do the following:
+
+1. Flash one node with openthread_ncp example with `make BOARD=<target> clean all flash`
```
add `-C tests/openthread_ncp`

biboc replied:
ok

aabadie commented on tests/openthread_ncp/README.md:
````diff
@@ -0,0 +1,43 @@
+1. Flash one node with openthread_ncp example with `make BOARD=<target> clean all flash`
+2. Install wpantund: On Ubuntu 16.04, run
+```
+sudo apt-get -y install build-essential subversion libncurses5-dev libssl-dev zlib1g-dev gawk gcc-multilib flex git-core gettext gcc binutils bzip2 python perl make unzip libz-dev tftp git shtool autogen automake libtool autotools-dev libdbus-1-dev
````
I think all this could be added in as a Makefile rule in `dist/tools/wpantund`, see uhcpd/ethos with gnrc_border_router example.
But you can leave it as is for this PR and do that in a follow-up PR.

biboc replied:
This will be in a following PR

aabadie commented on tests/openthread_ncp/README.md:
```diff
@@ -0,0 +1,43 @@
+OpenThread was ported. Please check the [full
+documentation](https://github.com/openthread/openthread/blob/master/src/cli/README.md) of the CLI for usage information.
+
+This test show how to use a Network Co-Processor (NCP). A Network Co-Processor is used with wpantund software.
```
show**s**

aabadie commented on tests/openthread_ncp/Makefile:
```diff
@@ -0,0 +1,58 @@
+# Comment this out to disable code in RIOT that does safety checking
+# which is not needed in a production environment but helps in the
+# development process:
+CFLAGS += -DDEVELHELP -Wall
```
this can be commented out by default

aabadie commented on tests/openthread_ncp/Makefile:
```diff
@@ -0,0 +1,58 @@
+# If no BOARD is found in the environment, use this default:
+BOARD ?= samr21-xpro
+
+# These are the board that OpenThread stack has been tested on
```
the board**s**

aabadie commented on tests/openthread_ncp/Makefile:
```diff
@@ -0,0 +1,58 @@
+USEMODULE += openthread_contrib
+USEMODULE += libmbedcrypto
+USEMODULE += libopenthread-ftd
+USEMODULE += libopenthread-ncp-ftd
```
I prefer using openthread-ftd and openthread-ncp as module names (e.g remove lib prefix). Other point: would it make sense to make openthread_contrib and libmbedcrypto dependencies of openthread ? Then USEPKG += openthread will import these minimum dependencies.

biboc replied:
I use libopenthread-ftd because this is the name of .a library:
libopenthread-ftd.a
libopenthread-ncp-ftd.a
So it makes sense for me to keep these name

Yes contrib et crypto could be dependencies of PKG openthread, where should I add PKG dependencies?
Makefile.dep
Something like:
```
ifneq (,$(filter openthread,$(USEPKG)))
USEMODULE += openthread_contrib
USEMODULE += libmbedcrypto
endif
```

aabadie commented on tests/openthread_mtd/main.c:
```diff
@@ -0,0 +1,23 @@
+
+int main(void)
+{
+    return 0;
```
Maybe add a welcome message ?

aabadie commented on tests/openthread_mtd/README.md:
```diff
@@ -0,0 +1,42 @@
+documentation](https://github.com/openthread/openthread/blob/master/src/cli/README.md) of the CLI for usage information.
+
+This test compiles a Minimal Thread Device (MTD). MTD does not have router functionality compiled in.
+As a result, it is not necessary to configure the routerrole on an MTD.
```
is routerrole expected or shoud it be router role ?

biboc replied:
no MTD can't be a router, I'll remove the line to avoid confusion

aabadie commented on pkg/openthread/include/platform_config.h:
```diff
@@ -0,0 +1,41 @@
+ * @}
+ */
+
+#ifndef OPENTHREAD_CORE_PLATFORM_CONFIG_RIOT_H_
```
trailing underscore

aabadie commented on pkg/openthread/include/platform_config.h:
```diff
@@ -0,0 +1,41 @@
+  */
+#define OPENTHREAD_CONFIG_LEGACY_TRANSMIT_DONE 1
+
+#endif /* OPENTHREAD_CORE_PLATFORM_CONFIG_RIOT_H_ */
```
trailing underscore. By the way, I think this header guard name won't pass the CI. I should be `PLATFORM_CONFIG_H`

aabadie commented on pkg/openthread/include/ot.h:
```diff
@@ -34,18 +35,46 @@ extern "C" {
 typedef struct {
-    void *buf;  /**< buffer containing the message */
-    size_t len; /**< length of the message */
+    uint16_t length;                             /**< length of the message */
```
please align doxygen

aabadie commented on pkg/openthread/include/ot.h:
```diff
@@ -34,18 +35,46 @@ extern "C" {
-    void *buf;  /**< buffer containing the message */
-    size_t len; /**< length of the message */
+    uint16_t length;                             /**< length of the message */
+    uint16_t serialBufferStatus;                 /**< status of the buffer */
```
use snake_case

aabadie commented on pkg/openthread/include/ot.h:
```diff
@@ -34,18 +35,46 @@ extern "C" {
+#define OPENTHREAD_JOB_MSG_TYPE_EVENT                       (0x2240)
+/**< number of serial reception buffer*/
+#define OPENTHREAD_NUMBER_OF_SERIAL_BUFFER                   1
+    /**< sizeof in bytes the two first members of she serial structure*/
```
extra spaces at beginning of line

aabadie commented on pkg/openthread/include/ot.h:
```diff
@@ -34,18 +35,46 @@ extern "C" {
+/**< event indicating an OT_JOB message */
+#define OPENTHREAD_JOB_MSG_TYPE_EVENT                       (0x2240)
+/**< number of serial reception buffer*/
+#define OPENTHREAD_NUMBER_OF_SERIAL_BUFFER                   1
```
parenthesis around value, use 1U

aabadie commented on pkg/openthread/include/ot.h:
```diff
@@ -34,18 +35,46 @@ extern "C" {
+/**< number of serial reception buffer*/
+#define OPENTHREAD_NUMBER_OF_SERIAL_BUFFER                   1
+    /**< sizeof in bytes the two first members of she serial structure*/
+#define OPENTHREAD_SIZEOF_LENGTH_AND_FREEBUFF                4
```
parenthesis, use 4U

aabadie commented on pkg/openthread/include/ot.h:
```diff
@@ -34,18 +35,46 @@ extern "C" {
+#define OPENTHREAD_SERIAL_BUFFER__PAYLOAD_SIZE               OPENTHREAD_SERIAL_BUFFER_SIZE - OPENTHREAD_SIZEOF_LENGTH_AND_FREEBUFF
+/**< error when no more buffer available*/
+#define OPENTHREAD_ERROR_NO_EMPTY_SERIAL_BUFFER              -1
+    /**< serial buffer ready to use*/
```
spaces at beginning of line

aabadie commented on pkg/openthread/contrib/platform_uart.c:
```diff
@@ -13,34 +13,148 @@
-    return kThreadError_None;
+    uint8_t i =0;
+
+    for(i=0; i < OPENTHREAD_NUMBER_OF_SERIAL_BUFFER; i++) {
```
add space after `for` + around operator `=`

aabadie commented on pkg/openthread/contrib/platform_uart.c:
```diff
@@ -13,34 +13,148 @@
+    msg_send_int(&msg, openthread_get_pid());
+}
+
+#endif //MODULE_LIBOPENTHREAD_NCP_FTD
```
don't use cpp style comments

aabadie commented on pkg/openthread/contrib/platform_random.c:
```diff
@@ -49,3 +49,14 @@ uint32_t otPlatRandomGet(void)
+otError otPlatRandomGetTrue(uint8_t *aOutput, uint16_t aOutputLength)
+{
+    for (uint16_t index = 0; index < aOutputLength; index++)
+    {
```
style:
```c
for (...) {
...
}
```

aabadie commented on pkg/openthread/contrib/platform_misc.c:
```diff
@@ -21,16 +21,24 @@
+
+void otPlatWakeHost(void)
+{
+    // TODO: implement an operation to wake the host from sleep state.
```
use c style comments: `/* */`

aabadie commented on pkg/openthread/contrib/platform_logging.c:
```diff
@@ -13,85 +13,34 @@
-#include <time.h>
 
-#include "openthread/platform/logging.h"
+#include <utils/code_utils.h>
```
use "" around includes.
`<>` are only for system includes

aabadie commented on pkg/openthread/contrib/platform_logging.c:
```diff
@@ -13,85 +13,34 @@
  * @}
  */
 
+#include <openthread/config.h>
```
"" around include

aabadie commented:
By the way, is this PR still WIP ? if not, you can remove WIP from the title.

biboc commented:
> I'm wondering if we could keep the mtd and ftd tests into a single test application to avoid code duplication.

No there are compiled the same way and behavior is different for each example

aabadie commented:
> No they are not compiled the same way and the behavior is different for each example

What I meant is that the `main.c` is very similar, we could simply rely on Makefile variables and decide at build time of a single application the kind of application:
```
# build a mtd:
make BOARD=samr21-xpro OPENTHREAD_TYPE=mtd -C tests/openthread flash term
# build a ftd:
make BOARD=samr21-xpro OPENTHREAD_TYPE=ftd -C tests/openthread flash term
# build a ncp:
make BOARD=samr21-xpro OPENTHREAD_TYPE=ncp -C tests/openthread flash term
```

biboc commented:
Ok good idea!

syin2 commented:
HI @biboc ,
I have cloned your branch and did a openthread_ftd test.
The board I'm using is remote-revb.
The CLI application does not work on my side. 
When I type help command, there is no responding msgs coming out to the stdout.
> help
2017-06-16 13:17:18,793 - INFO # 
2017-06-16 13:17:18,796 - INFO # 
2017-06-16 13:17:18,800 - INFO # 
2017-06-16 13:17:18,803 - INFO # 
2017-06-16 13:17:18,813 - INFO # 
2017-06-16 13:17:18,820 - INFO # 
Any idea why this might happen?

biboc commented:
@syin2, I don't have Zolertia RE-Mote board so I can't try, we've added a WHITE_LIST in the Makefile for all tested board
Moreover remote-revb use a TI's CC2538 and at the moment OpenThread has been tested only with at86rf2xx driver

biboc commented:
@aabadie, I choose to only merge FTD and MTD together and keep NCP since it is very different

biboc commented:
@aabadie, I addressed all your comments

miri64 commented on pkg/openthread/contrib/netdev/openthread_netdev.c:
```diff
@@ -98,32 +99,45 @@ static void *_openthread_event_loop(void *arg) {
-                break;
-        }
+        otTaskletsProcess(sInstance);
+         if (otTaskletsArePending(sInstance) == false) {
```
indentation is off here (at the closing bracket too)

biboc replied:
Done

aabadie commented on tests/openthread_ncp/Makefile:
```diff
@@ -0,0 +1,55 @@
+# interface driver/daemon that provides a native IPv6 network interface to a low-power wireless Network Co-Processor
+# FTD (see openthread_test_ftd): NCP needs Full Thread Device to act as a router in the network
+USEPKG += openthread
+USEMODULE += libopenthread-ftd
```
I prefer to change the module names to `openthread-ftd`, `openthread-ncp-ftd`, `openthread-posix` : easier to remember for users

aabadie commented on tests/openthread/Makefile:
```diff
@@ -11,43 +12,50 @@ RIOTBASE ?= $(CURDIR)/../..
 ifneq (,$(filter at86rf2%,$(DRIVER)))
-  FEATURES_REQUIRED = periph_spi
-  FEATURES_REQUIRED = periph_gpio
+  FEATURES_REQUIRED = periph_spi periph_gpio
```
FEATURES_REQUIRED **+**=

aabadie commented on tests/openthread/README.md:
```diff
@@ -1,36 +1,43 @@
+
+You can choose to compule a FTD or MTD with :
+`make BOARD=<target> clean all flash OPENTHREAD_TYPE=mtd/ftd`
+- Minimal Thread Device (MTD) does not have router functionality compiled in.An MTD may or may not be sleepy.
```
Missing space after `.` and before 'An MTD'

aabadie commented on tests/openthread/README.md:
```diff
@@ -1,36 +1,43 @@
 
+To test OpenThread in RIOT, you can do the following:
+
+1. Flash nodes with `make BOARD=<target> clean all flash OPENTHREAD_TYPE=mtd/ftd`
```
Split this in 2 lines: one for each type of build (mtd and ftd). I can imagine brainless users try the exact command and falling the ftd case.

aabadie commented on tests/openthread/README.md:
```diff
@@ -1,36 +1,43 @@
+To test OpenThread in RIOT, you can do the following:
+
+1. Flash nodes with `make BOARD=<target> clean all flash OPENTHREAD_TYPE=mtd/ftd`
+2. Check the state of the node with `state`. In the beggining, it should be `detached`, but after some seconds it should
```
beginning

aabadie commented on tests/openthread/Makefile:
```diff
@@ -11,43 +12,50 @@ RIOTBASE ?= $(CURDIR)/../..
+OPENTHREAD_TYPE ?= ftd
+ifeq ($(OPENTHREAD_TYPE),mtd)
+    # MTD: A Minimal Thread Device does not have router functionality compiled in.
+    # As a result, it is not necessary to configure the routerrole on an MTD. At the same time, an MTD may or may not be sleepy.
```
split this line to keep it < 80 characters wide

biboc replied:
I'll do it but today, computer screen are much larger 120 bytes is good also ;-)

aabadie commented on pkg/openthread/include/ot.h:
```diff
@@ -34,18 +35,46 @@ extern "C" {
+#define OPENTHREAD_JOB_MSG_TYPE_EVENT                       (0x2240)
+/**< number of serial reception buffer*/
+#define OPENTHREAD_NUMBER_OF_SERIAL_BUFFER                   (1U)
+    /**< sizeof in bytes the two first members of she serial structure*/
```
extra spaces

aabadie commented on pkg/openthread/contrib/platform_uart.c:
```diff
@@ -13,34 +13,148 @@
+otError otPlatUartEnable(void)
 {
-    return kThreadError_None;
+    uint8_t i =0;
```
`i` should be declared in the `for ()`, use insigned:
```c
for (unsigned i = 0; i < ...) {
...
}
```

aabadie commented on pkg/openthread/contrib/platform_uart.c:
```diff
@@ -13,34 +13,148 @@
+        if ( OPENTHREAD_ERROR_NO_EMPTY_SERIAL_BUFFER == currentSerialBufferNumber ) {
+            return;
+        }
+        if( OPENTHREAD_SERIAL_BUFFER_STATUS_FULL != gSerialMessage[currentSerialBufferNumber]->serial_buffer_status) {
```
missing space after if

aabadie commented on pkg/openthread/contrib/platform_uart.c:
```diff
@@ -13,34 +13,148 @@
+
+#ifdef MODULE_LIBOPENTHREAD_NCP_FTD
+int8_t getFirstEmptySerialBuffer(void) {
+    uint8_t i =0;
```
move declaration in for, use unsigned

biboc replied:
Ok to move but why use unsignerd, the number of buffer will never goes over 255 so I keep uint8_t

aabadie commented on pkg/openthread/contrib/platform_uart.c:
```diff
@@ -13,34 +13,148 @@
+
+#define UART_OPENTHREAD_BAUDRATE            (115200U)
+#define UART_OPENTHREAD                     UART_DEV(0)
+#define OPENTHREAD_SPINEL_FRAME_MARKER      0x7e
```
`()` around value

aabadie commented on pkg/openthread/contrib/platform_logging.c:
```diff
@@ -13,85 +13,33 @@
+    (void)aLogLevel;
+    (void)aLogRegion;
+    (void)aFormat;
+#endif // OPENTHREAD_ENABLE_CLI_LOGGING
```
cpp style comment, use `/* */` instead

aabadie commented on pkg/openthread/contrib/netdev/openthread_netdev.c:
```diff
@@ -83,12 +88,8 @@ static void *_openthread_event_loop(void *arg) {
     netdev_t *dev;
     msg_t msg, reply;
 
+#if defined(MODULE_LIBOPENTHREAD_CLI_FTD) || defined(MODULE_LIBOPENTHREAD_CLI_MTD)
```
I don't understand how and where modules LIBOPENTHREAD_CLI_FTD and LIBOPENTHREAD_CLI_MTD are defined ?

biboc replied:
https://github.com/RIOT-OS/RIOT/pull/7149/files#diff-b7c878ba1ad0b6981cab03da75f2b1bdR26

Each USEMODULE added is defined as MODULE_<NAME_OF_MODULE>, I made use of that

biboc commented:
@aabadie, I understand your point of view about libopenthread-cli-ftd but it is more understandable for me to link the .a library and the RIOT MODULE:

In the OpenThread Makefile:
```
ifneq (,$(filter libopenthread-cli-ftd,$(USEMODULE)))
	cp $(PKG_BUILDDIR)/output/lib/libopenthread-ftd.a ${BINDIR}/libopenthread-ftd.a
	cp $(PKG_BUILDDIR)/output/lib/libopenthread-cli-ftd.a ${BINDIR}/libopenthread-cli-ftd.a
endif
```

Otherwise I need to cp and rename openthread lib just to remove the `lib` before `libopenthread-clid-ftd`

biboc commented:
Ok @aabadie, I moved all libopenthread-xxx to openthread-xxx

aabadie commented on tests/openthread/Makefile:
```diff
@@ -11,43 +12,47 @@ RIOTBASE ?= $(CURDIR)/../..
   DRIVER := at86rf233
 endif
-ifneq (,$(filter iotlab-m3 fox iotlab-a8-m3,$(BOARD)))
+ifneq (,$(filter iotlab-m3 fox,$(BOARD)))
```
why removing `iotlab-a8-m3` ? This 'board' is very similar to iotlab-m3 and should be kept.

biboc replied:
Done

aabadie commented on tests/openthread/Makefile:
```diff
@@ -11,43 +12,47 @@ RIOTBASE ?= $(CURDIR)/../..
+endif
+
+#Define PANID and CHANNEL used by default
+#CFLAGS += -DOPENTHREAD_PANID=0xbeef -DOPENTHREAD_CHANNEL=11
```
I think it would be even simpler to use special variables for each with default values:
```make
OPENTHREAD_PANID ?= 0xbeef
OPENTHREAD_CHANNEL ?= 11

CFLAGS += -DOPENTHREAD_PANID=${OPENTHREAD_PANID} -DOPENTHREAD_CHANNEL=${OPENTHREAD_CHANNEL}
```
Then user could pass the variable independently when building the firmware and without having to edit the Makefile
```
make BOARD=samr21-xpro OPENTHREAD_PANID=0xaaaa OPENTHREAD_TYPE=ftd flash term
```

biboc replied:
Good one!

aabadie commented on tests/openthread/README.md:
```diff
@@ -1,36 +1,45 @@
 
-This test demonstrates the [OpenThread](https://github.com/openthread/openthread) stack running on RIOT. When flashed,
-it will initialize the OpenThread Command Line Interface for interacting with the stack.
+This test demonstrates how to use OpenThread stack from [OpenThread](https://github.com/openthread/openthread) on
```
I would rewrite the first phrase like this:
This test demonstrates how to use the [OpenThread](https://github.com/openthread/openthread) open source implementation of [Thread](https://threadgroup.org/).

aabadie commented on tests/openthread/README.md:
```diff
@@ -1,36 +1,45 @@
+OpenThread was ported. Please check the [full
+documentation](https://github.com/openthread/openthread/blob/master/src/cli/README.md) of the CLI for usage information.
+
+You can choose to compule a FTD or MTD with :
```
You can either build a FTD or MTD firmware with:

aabadie commented on tests/openthread/README.md:
```diff
@@ -1,36 +1,45 @@
 
-To test OpenThread on RIOT, you can do the following:
+With RIOT port, a node is auto-setup and ready to communicate with this configuration:
+PANID: 0x1234
```
This may change with the suggestion above.

aabadie commented on tests/openthread/main.c:
```diff
@@ -10,25 +10,21 @@
     uint8_t res = ot_call_command("panid", NULL, (void*)&panid);
     printf("Current panid: 0x%x (res:%x)\n", panid, res);
 
-    openthread_uart_run();
```
Are you sure the CLI is still available with this line removed ?

biboc replied:
Yes I'm sure

aabadie commented on tests/openthread_ncp/Makefile:
```diff
@@ -0,0 +1,51 @@
+ifneq (,$(filter samr21-xpro,$(BOARD)))
+  DRIVER := at86rf233
+endif
+ifneq (,$(filter iotlab-m3 fox,$(BOARD)))
```
iotlab-a8-m3 to be added here: very important if someone wants to test on IoT-LAB. This is the only board that can be used as an NCP (for the moment) on the testbed.

aabadie commented on tests/openthread_ncp/Makefile:
```diff
@@ -0,0 +1,51 @@
+endif
+
+ifneq (,$(filter at86rf2%,$(DRIVER)))
+  FEATURES_REQUIRED = periph_spi periph_gpio
```
`+=`

aabadie commented on tests/openthread_ncp/README.md:
```diff
@@ -0,0 +1,45 @@
+## OpenThread on RIOT
+
+This test demonstrates how to use OpenThread stack from [OpenThread](https://github.com/openthread/openthread) on
```
This is copy pasted from openthread test. Might not be very useful, you can link to the other README file directly (See [openthread test](../openthread/README.md) for more detail).

aabadie commented on tests/openthread_ncp/main.c:
```diff
@@ -0,0 +1,24 @@
+
+int main(void)
+{
+    /* Run wpantund to interact with NCP */
```
Maybe add a welcome message with some information here (using puts).

biboc replied:
openthread_ncp aims to be connected to wpantund so the welcome message appears when you run it and wpantund is connected to by UART, No need to write something in RIOT main

biboc commented:
@aabadie, made your change

aabadie commented:
Thanks @biboc for addressing the comments quickly. I think that the changes are good codewise. What remain are:
- testing on IoT-Lab or any other local setup
- squashing the commits (I think 2 or 3 are needed only)

jia200x commented:
The `thread start` command hangs the CLI shell if I change the channel. I'm investigating why

biboc commented:
@aabadie Agree! Please could you test in IoT-Lab? I've never done it.
Squashing, yes sure

@jia200x, #7207 fix the issue ;-)

biboc commented:
@jia200x #7115 solves the channel problem. 
I'm going to squash my commits

biboc commented:
@aabadie, I've done only one commit since all edits are dependent. 
I can't leave old openthread example and only commit change in contrib folder otherwise previous example won't work. 
Only thing I can do is two commits: one with the entire port and openthread example and one with openthread_ncp example

aabadie commented:
> Only thing I can do is two commits: one with the entire port and openthread example and one with openthread_ncp example

I buy it !

biboc commented:
Done, let's wait murdrock

aabadie commented:
@biboc can you amend the commit messages to match the RIOT rules:
* pkg/openthread: add FTD and NCP support
* tests: add openthread NCP application

Or something like this

biboc commented:
Done

biboc commented:
@aabadie I let you dismiss your requested changes and let's go!

aabadie commented:
Let's wait for #7115 to be merged first

miri64 commented:
(restarted build to show the proper warning ;-))

jia200x commented:
When this is merged I will rebase my OpenThread sock adoption an push another PR :)

thomaseichinger commented:
@aabadie @biboc FYI, #7115 is merged.

biboc commented:
@aabadie, did you try the PR? 
Let's merge?

aabadie commented:
I wish I could, will test tomorrow. I hope I'll procrastinate less.

aabadie commented:
What is your setup @biboc ?

aabadie commented:
the master version works by the way (and I can ping nodes with openthread cli).

aabadie commented:
updated [comment](https://github.com/RIOT-OS/RIOT/pull/7149#pullrequestreview-47062344)

biboc commented:
@aabadie I use a samr21-xpro connected by USB to an Ubuntu and it works.

> The NCP cannot be commissioned and no wpan0 interface is created on the A8

wpan0 is created by wpantund, not the NCP

May I try on IoT-LAB? I've never done that?

> The NCP cannot be commissioned 

What do you mean by commissioned?

biboc commented:
I've improved the CLI handler so it read an entire command before sending it to OpenThread.
It works for samr21-xpro.
Tell me how it works on IoT-LAB

miri64 commented:
@biboc with #7275 merged, does this still work?

biboc commented:
I should rebase on master to see this

miri64 commented:
For testing you also could just merge current master into your branch. ;-)

aabadie commented:
@biboc I tested on IoT-Lab again. It works with samr21-xpro but not with iotlab-m3.

I also noticed a problem with samr21-xpro: the command are cropped when passed to OpenThread CLI. Here is the output of some commands:
```
> panid
panid
beef
Done
> state
state
router
Done
> version
versi
Error 6: Parse
> ipaddr
ipadd
Error 6: Parse
```
I cannot get the IP of a node: ipaddr is cropped...

aabadie commented:
If you want to test on IoT-LAB, you can [create an account](https://www.iot-lab.info/testbed/signup.php), once validated by the admins, there are [good tutorials available](https://www.iot-lab.info/tutorials/) to start using the platform.

biboc commented:
@aabadie, I've just pushed a commit with UART improved and 2.7KO RAM saved!

aabadie commented:
@biboc, kernel panic :)

```
main(): This is RIOT! (Version: 2017.04-devel-1473-g8f0ef-trek-saclay-openthread-FTD-MTD-NCP_V2)
This a test for OpenThread
Get PANID�Current panid: 0xbeef (res:0)
help
*** RIOT kernel panic:
HARD FAULT HANDLER

*** rebooting...�main(): This is RIOT! (Version: 2017.04-devel-1473-g8f0ef-trek-saclay-openthread-FTD-MTD-NCP_V2)
This a test for OpenThread
Get PANID�Current panid: 0xbeef (res:0)
```

biboc commented:
Yes pull lastest changed please

biboc commented:
It works on my side (samr21-xpro):

```
2017-06-29 15:47:23,672 - INFO #  panid
2017-06-29 15:47:23,673 - INFO # beef
2017-06-29 15:47:23,673 - INFO # Done
> state
2017-06-29 15:47:25,052 - INFO #  state
2017-06-29 15:47:25,053 - INFO # leader
2017-06-29 15:47:25,054 - INFO # Done
> version
2017-06-29 15:47:31,041 - INFO #  version
2017-06-29 15:47:31,045 - INFO # OPENTHREAD/g8dc31d6; none; Jun 29 2017 14:48:50
2017-06-29 15:47:31,045 - INFO # Done
> ipaddr
2017-06-29 15:47:34,531 - INFO #  ipaddr
2017-06-29 15:47:34,534 - INFO # fdde:ad00:beef:0:0:ff:fe00:fc00
2017-06-29 15:47:34,537 - INFO # fdde:ad00:beef:0:0:ff:fe00:2c00
2017-06-29 15:47:34,540 - INFO # fdde:ad00:beef:0:3c:e737:feaa:6628
2017-06-29 15:47:34,543 - INFO # fe80:0:0:0:b8ad:4c41:2ac5:9290
2017-06-29 15:47:34,544 - INFO # Done
```

aabadie commented:
weird, still have the kernel panic. I'm on f77741219ce39235b2d6ef99daff766a6a127a07

biboc commented:
I replaced the last commit so try to pull again and check this in ot.h:
```
#define OPENTHREAD_SERIAL_BUFFER_SIZE                       OPENTHREAD_SIZEOF_LENGTH_AND_FREEBUFF + 100

```

biboc commented:
@aabadie better?

aabadie commented:
> @aabadie better?

Sorry, I couldn't test again (working on #6797). I'll give another try tomorrow morning.

aabadie commented:
@biboc, I still have the kernel panic on IoT-Lab with samr21-xpro and shell doesn't work with iotlab-m3. I suspect a baudrate issue with iotlab-m3 since they use 500000 for stdio.

For samr21-xpro I suspect a timeout with flasher script since the firmware is very big (text: 160488 byte).

aabadie commented:
Tested locally on samr21-xpro: still kernel panic. Sorry, I don't have the time to look deeper into that issue @biboc.

biboc commented:
@aabadie, Do it this way and tell me if it works (works for me): 

```
@mkdir tmp && cd tmp
@git clone https://github.com/biboc/RIOT.git
@cd RIOT
@git checkout checkout openthread-FTD-MTD-NCP_V2
@cd tests/openthread/
@make flash term BOARD=samr21-xpro
Building application "openthread" for "samr21-xpro" with MCU "samd21".

"make" -C /home/baptiste/EISOX/workspace_valve/tmp/RIOT/pkg/openthread
Compile OpenThread for FTD device
$OPENTHREAD_ARGS is [--enable-cli-app=ftd --enable-application-coap]

....

"make" -C /home/baptiste/EISOX/workspace_valve/tmp/RIOT/sys/tsrb
"make" -C /home/baptiste/EISOX/workspace_valve/tmp/RIOT/sys/uart_stdio
"make" -C /home/baptiste/EISOX/workspace_valve/tmp/RIOT/sys/xtimer
   text	   data	    bss	    dec	    hex	filename
 161288	    512	  19080	 180880	  2c290	/home/baptiste/EISOX/workspace_valve/tmp/RIOT/tests/openthread/bin/samr21-xpro/openthread.elf
/home/baptiste/EISOX/workspace_valve/tmp/RIOT/dist/tools/edbg/edbg  -t atmel_cm0p -b -e -v -p -f /home/baptiste/EISOX/workspace_valve/tmp/RIOT/tests/openthread/bin/samr21-xpro/openthread.bin
Debugger: ATMEL EDBG CMSIS-DAP ATML2127031800002954 03.0C.0191 (S)
Target: SAM R21G18
Erasing...  done.
Programming............................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................ done.
Verification............................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................ done.
/home/baptiste/EISOX/workspace_valve/tmp/RIOT/dist/tools/pyterm/pyterm -p "/dev/ttyACM0" -b "115200"
No handlers could be found for logger "root"
2017-06-30 11:10:55,897 - INFO # Connect to serial port /dev/ttyACM0
Welcome to pyterm!
Type '/exit' to exit.
state
2017-06-30 11:11:00,122 - INFO # state
2017-06-30 11:11:00,125 - INFO # leader
2017-06-30 11:11:00,126 - INFO # Done
> panid
2017-06-30 11:11:02,474 - INFO #  panid
2017-06-30 11:11:02,475 - INFO # beef
2017-06-30 11:11:02,476 - INFO # Done
> ipaddr
2017-06-30 11:11:04,315 - INFO #  ipaddr
2017-06-30 11:11:04,318 - INFO # fdde:ad00:beef:0:0:ff:fe00:fc00
2017-06-30 11:11:04,321 - INFO # fdde:ad00:beef:0:0:ff:fe00:8c00
2017-06-30 11:11:04,325 - INFO # fdde:ad00:beef:0:4cc2:aa42:4326:13a3
2017-06-30 11:11:04,328 - INFO # fe80:0:0:0:4c2d:95c0:acf1:40a0
2017-06-30 11:11:04,328 - INFO # Don
```

aabadie commented:
@biboc, this is very weird, I'm on the latest commit and I flash locally like you do but I still get the kernel panic.

aabadie commented on pkg/openthread/contrib/platform_uart.c:
```diff
@@ -13,34 +13,164 @@
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+#define UART_OPENTHREAD_BAUDRATE            (115200U)
```
This won't work on iotlab-m3 since it works at 500000. You should add this define in the openthread configuration file. Same for `UART_OPENTHREAD`.

biboc replied:
ok

aabadie commented on pkg/openthread/contrib/platform_uart.c:
```diff
@@ -13,34 +13,164 @@
+#include "debug.h"
+
+#define UART_OPENTHREAD_BAUDRATE            (115200U)
+#define UART_OPENTHREAD                     UART_DEV(0)
```
I would call this `OPENTHREAD_UART_DEV`

biboc replied:
Yes I'll do

biboc commented:
@aabadie done, it should work on both boards now

Please git clone in new directory my repository and checkout to PR branch before trying again

emmanuelsearch commented:
not a priority to fix, but building fails on OS X with

```
/Users/emmanuel/Documents/work/CODING/RIOT/tests/openthread/bin/pkg/samr21-xpro/openthread/third_party/nlbuild-autotools/repo/scripts/bootstrap: line 180: /Users/emmanuel/Documents/work/CODING/RIOT/tests/openthread/bin/pkg/samr21-xpro/openthread/third_party/nlbuild-autotools/repo/tools/host/i386-apple-darwin/bin/libtoolize: No such file or directory
chmod: /tmp/tmp.bootstrap0bVJnI/libtoolize: No such file or directory
/Users/emmanuel/Documents/work/CODING/RIOT/tests/openthread/bin/pkg/samr21-xpro/openthread/third_party/nlbuild-autotools/repo/scripts/bootstrap: line 239: /tmp/tmp.bootstrap0bVJnI/libtoolize: No such file or directory
cd /Users/emmanuel/Documents/work/CODING/RIOT/tests/openthread/bin/pkg/samr21-xpro/openthread && CPP="arm-none-eabi-gcc -E" CC="arm-none-eabi-gcc" CXX="arm-none-eabi-g++"\
		OBJC="" OBJCXX="" AR="arm-none-eabi-ar" RANLIB="" NM="" \
		STRIP="" \
		CPPFLAGS="-fdata-sections -ffunction-sections -Os -mcpu=cortex-m0plus -mlittle-endian -mthumb -mfloat-abi=soft -mno-thumb-interwork -DOPENTHREAD_PROJECT_CORE_CONFIG_FILE='\"platform_config.h\"'" \
		CFLAGS="-fdata-sections -ffunction-sections -Os -mcpu=cortex-m0plus -mlittle-endian -mthumb -mfloat-abi=soft -mno-thumb-interwork " \
		CXXFLAGS="-fdata-sections -ffunction-sections -Os -mcpu=cortex-m0plus -mlittle-endian -mthumb -mfloat-abi=soft -mno-thumb-interwork -fno-exceptions -fno-rtti " \
		LDFLAGS="-fdata-sections -ffunction-sections -Os -mcpu=cortex-m0plus -mlittle-endian -mthumb -mfloat-abi=soft -mno-thumb-interwork -nostartfiles -specs=nano.specs \
		-specs=nosys.specs -Wl,--gc-sections -Wl,-Map=map.map " \
		./configure --disable-docs --host=arm-none-eabi --target=arm-none-eabi \
		--prefix=/ --enable-default-logging --enable-cli-app=mtd --enable-joiner --enable-application-coap
/bin/sh: ./configure: No such file or directory
make[1]: *** [all] Error 127
make: *** [/Users/emmanuel/Documents/work/CODING/RIOT/tests/openthread/bin/samr21-xpro/openthread.a] Error 2
```

aabadie commented:
@emmanuelsearch it looks like you are missing `libtoolize`. But that definitely an issue, do you also have it with `master` ?

aabadie commented:
Sorry, I couldn't test again this one but the changes look correct now. Unfortunately I have to postpone because of feature freeze. 

BTW, thanks for the efforts @biboc.

Hyungsin commented:
Glad to see this PR 👍  
I am considering using RIOT/OpenThread for the network stack of our hamilton (SAMR21-based platform with 3~7 sensors, idle current~=8uA and $10~20) project. I'll be more than happy if this PR ends soon! 
I'm also down to try IoT-LAB :))

biboc commented:
@aabadie, feature freeze? I made effort to merge this PR in master for July release. Only thing to do is to try on Iot-Lab. @Hyungsin can you do that?
Otherwise I remove @iot-lab board from the PR and we can merge it.

biboc commented:
@aabadie, I told you that I ammend the last commit so I'm sure pulling will update the last commit since it is the same one. Please create a new RIOT clone and it will work, so let's merge this PR for July release

aabadie commented:
> Please create a new RIOT clone and it will work, so let's merge this PR for July release

That's what I did. I also retested today on IoT-LAB and here are the results:
* samr21-xpro and iotlab-m3 work when the firmware is built on the SSH frontend (arm-none-eabi-gcc 4.9.3)
* samr21-xpro and iotlab-m3 don't work when the firmware is built on my laptop (arm-none-eabi-gcc 6.3.1). Samr21-xpro crashes and iotlab-m3 shell output is broken. With master, I don't have this problem.
* The NCP doesn't work with the wpantund  built on the A8 linux image (Yocto), wpantund is in version  0.07.01 (`0696fea`). Here is the output when I start it on the A8:
```
root@node-a8-141:~# /usr/sbin/wpantund -s /dev/ttyA8_M3 -b 500000 -I wpan0 -o SyslogMask all
wpantund[1941]: Starting wpantund 0.07.01 (Jun 28 2017 21:05:02) . . .
wpantund[1941]: 	SOURCE_VERSION = 0.07.01
wpantund[1941]: 	BUILD_VERSION = 0696fea
wpantund[1941]: Reading configuration from "/etc/wpantund.conf" . . .
wpantund[1941]: Configuration file "/etc/wpantund.conf" read.
wpantund[1941]: CONFIGURATION PROPERTY "SyslogMask" IS DEPRECATED. Please use "Daemon:SyslogMask" instead.
wpantund[1941]: set-config-param: "Config:NCP:SocketBaud" = "500000"
wpantund[1941]: set-config-param: "Config:NCP:SocketBaud" set succeded
wpantund[1941]: set-config-param: "Config:NCP:SocketPath" = "/dev/ttyA8_M3"
wpantund[1941]: set-config-param: "Config:TUN:InterfaceName" = "wpan0"
wpantund[1941]: set-config-param: "Daemon:SyslogMask" = "all"
wpantund[1941]: set-config-param: "Daemon:SyslogMask" set succeded
wpantund[1941]: Getting DBus connection
wpantund[1941]: Registering DBus connection
wpantund[1941]: Requesting DBus connection name com.nestlabs.WPANTunnelDriver
wpantund[1941]: Ready. Using DBUS bus ":1.9"
wpantund[1941]: Stat (autolog): Tx: 0 packets (0 udp, 0 tcp, 0 icmp6) -- 0 bytes
wpantund[1941]: Stat (autolog): Tx history is empty
wpantund[1941]: Stat (autolog): 
wpantund[1941]: Stat (autolog): Rx: 0 packets (0 udp, 0 tcp, 0 icmp6) -- 0 bytes
wpantund[1941]: Stat (autolog): Rx history is empty
wpantund[1941]: Stat (autolog): 
wpantund[1941]: Stat (autolog): NCP state history is empty.
wpantund[1941]: Stat (autolog): 
wpantund[1941]: Stat (autolog): 
wpantund[1941]: Opening tun interface socket with name "wpan0"
wpantund[1941]: Running as root without dropping privileges!
wpantund[1941]: SELECT: fd_count=8 cms_timeout=0
wpantund[1941]: SELECT:     read_fd_set: 4, 7
wpantund[1941]: SELECT:     write_fd_set: 
wpantund[1941]: Initializing NCP
wpantund[1941]: NCP is now BUSY.
wpantund[1941]: SELECT: fd_count=8 cms_timeout=0
wpantund[1941]: SELECT:     read_fd_set: 4, 5, 6, 7
wpantund[1941]: SELECT:     write_fd_set: 
wpantund[1941]: SELECT: fd_count=8 cms_timeout=100
wpantund[1941]: SELECT:     read_fd_set: 4, 5, 6, 7
wpantund[1941]: SELECT:     write_fd_set: 
wpantund[1941]: [->NCP] CMD_PROP_VALUE_GET(PROP_PROTOCOL_VERSION) tid:1
wpantund[1941]: 	↳ 810201
wpantund[1941]: SELECT: fd_count=8 cms_timeout=5000
wpantund[1941]: SELECT:     read_fd_set: 4, 5, 6, 7
wpantund[1941]: SELECT:     write_fd_set: 
wpantund[1941]: ../../../git/src/ncp-spinel/SpinelNCPInstance-Protothreads.cpp:436: Requirement Failed (IS_EVENT_FROM_NCP(event) && GetInstance(this)->mInboundHeader == mLastHeader)
```
No wpan0 interface is created. I don't know/understand what happens. I didn't try it locally yet with a samr21-xpro but I guess it should work (if I don't build the firmware with gcc 6.3.1).

To conclude, I think this PR is not ready/tested enough to justify it to be part of a release. Postponing will give us more time to strengthen it.

>  I made effort to merge this PR in master for July release

I know and I really appreciate that.

It took a difficult decision with postponing this PR but I had to take it to ensure the release can arrive on time.

Anyway, this doesn't prevent us to continue to work (hard!) on it and merge it soon: feature freeze doesn't mean nothing can be merged into master. It just mean new features merged into master won't be back-ported into the release branch.

biboc commented:
Ok, we're both maintainer but your are assignee to this PR so you get the final choice.
My compiler: `arm-none-eabi-gcc version 5.4.1 20160919 (release)`

> wpantund is in version 0.07.01

It should be fine.

I can't do anything since it works fine on my computer+board so @aabadie I leave the PR testing in your hands.

biboc commented:
@aabadie, any chance you have try to flash again?

biboc commented:
@aabadie, you told me:
> samr21-xpro and iotlab-m3 don't work when the firmware is built on my laptop (arm-none-eabi-gcc 6.3.1). Samr21-xpro crashes and iotlab-m3 shell output is broken. With master, I don't have this problem.

I updated gcc and I get similar problem, I've fixed it. Could you try again please?

aabadie commented:
> I updated gcc and I get similar problem, I've fixed it.

Very nice

> Could you try again please?

I just did on IoT-Lab:
* works on samr21-xpro and iotlab-m3
* doesn't work on iotlab-a8-m3 and I don't understand why. It looks like an UART issue (baudrate maybe) but normally I passed the right value (500000, the same that I use for iotlab-m3, which works)

The problem with the iotlab-a8-m3 would also explain why the NCP cannot be commissionned (e.g initialized) since everything is done via UART.

I'll test with some local hardware later (I don't have any with me now).

aabadie commented on pkg/openthread/contrib/openthread.c:
```diff
@@ -76,5 +57,5 @@ void openthread_bootstrap(void)
 
     openthread_radio_init(netdev, tx_buf, rx_buf);
-    openthread_netdev_init(ot_thread_stack, sizeof(ot_thread_stack), THREAD_PRIORITY_MAIN - 5, "openthread", netdev);
+    openthread_netdev_init(ot_thread_stack, sizeof(ot_thread_stack), THREAD_PRIORITY_MAIN + 1, "openthread", netdev);
```
Why changing thread priority ?

aabadie commented on pkg/openthread/include/platform_config.h:
```diff
@@ -0,0 +1,51 @@
+ * The number of message buffers in buffer pool
+ */
+#if OPENTHREAD_ENABLE_NCP_UART
+#define OPENTHREAD_CONFIG_NUM_MESSAGE_BUFFERS       5
```
Put () around values

Hyungsin replied:
This buffer size is so mall, which is the main reason why we cannot ping through the border router. I increased it to 40 and ping works well now :)

aabadie replied:
The NCP doesn't build on samr21-xpro (not enough RAM) with this value, I decreased it to 30. The actual branch of this PR has been rebased on latest master.

aabadie commented on tests/openthread/Makefile:
```diff
@@ -11,17 +12,33 @@ RIOTBASE ?= $(CURDIR)/../..
 # which is not needed in a production environment but helps in the
 # development process:
-CFLAGS += -DDEVELHELP -Wall
+#CFLAGS += -DDEVELHELP -Wall
```
Should be reverted otherwise it's not any more in sync with the comment above.

aabadie commented on pkg/openthread/contrib/platform_uart.c:
```diff
@@ -13,34 +13,163 @@
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+#define OPENTHREAD_UART_DEV                 UART_DEV(0)
```
I would prefer setting UART_STDIO_BAUDRATE instead of UART_DEV(0)

aabadie commented on pkg/openthread/contrib/platform_uart.c:
```diff
@@ -13,34 +13,163 @@
+        gSerialMessage[i]->serial_buffer_status = OPENTHREAD_SERIAL_BUFFER_STATUS_FREE;
+    }
+
+    uart_init(OPENTHREAD_UART_DEV, OPENTHREAD_UART_BAUDRATE, (uart_rx_cb_t) uart_handler, NULL);
```
And here use UART_STDIO_DEV and UART_STDIO_BAUDRATE. Then you use the value defined in the board config and there's no need for passing the baudrate in the application makefile.

aabadie commented:
@biboc, I gave another try to the NCP. Not on IoT-LAB but locally with 2 samr21 and I confirm that it works. Well I can play with wpanctl interactively, start a network, have the wpan0, etc
But I couldn't ping the nodes in the network behind from my laptop. Maybe I missed something with the network configuration.
This makes me think that the test application readme could be a bit more detailed/improved.

aabadie commented on tests/openthread_ncp/README.md:
```diff
@@ -0,0 +1,43 @@
+./bootstrap.sh
+./configure
+make
+sudo make install
```
I'm not a big fan of this kind of command line. Better use the `--prefix` of the configure option and pass the configuration file from the `wpantund` command line (use `-c` option)

aabadie commented on tests/openthread_ncp/README.md:
```diff
@@ -0,0 +1,43 @@
+
+`sudo /usr/local/bin/wpanctl -I wpan0`
+
+5. With wpanctl CLI, you can use "status" cmd to check status of NCP, "scan" cmd scan for WPAN network around or "form" to create new WPAN network
```
This could be a bit more detailed I think

aabadie commented on tests/openthread_ncp/README.md:
````diff
@@ -0,0 +1,43 @@
+2. Install wpantund: On Ubuntu 16.04, run
+```
+sudo apt-get -y install build-essential subversion libncurses5-dev libssl-dev zlib1g-dev gawk gcc-multilib flex git-core gettext gcc binutils bzip2 python perl make unzip libz-dev tftp git shtool autogen automake libtool autotools-dev libdbus-1-dev
+git clone https://github.com/openthread/wpantund.git
````
Not sure of this but maybe giving a fixed commit hash will avoid problems when building the tool: we never know, master could be broken or not compatible with package version in RIOT.

biboc commented:
@aabadie, unfortunately, I won't work on OpenThread for a while so feel free, to add PR to my branch or create new PR based on this one

Hyungsin commented:
@aabadie and @biboc , please change the buffer size of ncp (platform_config.h, I added a comment there). The small buffer size blocks ping through the border router. At least in my case, ping works once the buffer size is increased :)

aabadie commented:
@biboc, @Hyungsin, I fixed a couple of things (force pushed in the branch directly) and tested again by PR:
* remaining style issues
* use default UART stdio and baudrate for OT shell and NCP control
* increased the NCP num message buffers to 30 because of RAM issues on sam21-xpro
* the NCP works for me: I can control it with `wpanctl`

Can you test it again with the latest changes and rebase @Hyungsin ?

Hyungsin commented:
@aabadie, sorry, I'm busy working on hamilton software upload. 
It would be good if any of you @biboc and @aabadie can test and finish this PR.

I can start another openthread-related PR after this is merged, adding some features and supporting the latest openthread code.

biboc commented:
Sorry I have no time for the moment to work on this PR

jia200x commented:
@Hyungsin Did you already start with the latest openthread code? I was planning to do so, but in case you are into that I can help with reviewing/testing.

Hyungsin commented:
@jia200x. Yes, I'm working on that and will open a PR after this PR is merged.
It would be great if you can help with reviewing/testing! 
So can you first test this PR to be merged?

aabadie commented:
> So can you first test this PR to be merged?

I just pushed an updated version of this PR. Now it's ok code wise but it just need to be tested before merging. Some things to check:
* FTD nodes can ping/coap each other
* can configure the NCP
* can ping a FTD node via the NCP (which is what this PR is meant for)

Hyungsin commented:
@aabadie cool! @jia200x, can you test this?

aabadie commented:
@Hyungsin, @jia200x, it would be very cool if you could test this PR soon with the 3 cases described above. Thanks !

jia200x commented:
Hi @aabadie, @Hyungsin 

I've been very busy these weeks, but I will try to find some time.

aabadie commented:
@jia200x @Hyungsin any chance for you to test this one ? (it needs a rebase btw)

jia200x commented:
@aabadie @Hyungsin Not yet, I will work on it this Monday - Tuesday. I will let you know how the tests go

jia200x commented:
FTD is not working for me. I'm investigating what's going on.

jia200x commented:
FTD is broken because OpenThread is calling otPlatRadioSleep before otPlatRadioReceive, and the state machine of at86rf2xx hangs because of this (fixed in #8186 5 days ago).

So, I will need to rebase before testing. I will continue tomorrow.

jia200x commented:
OK, I gave it a try now ;).
After rebase, ping/coap are working in FTD. Also, other commands like 'scan' work perfect.

Tomorrow I will try NCP configuration and ping via NCP

aabadie commented:
Thanks a lot for testing @jia200x. I also rebased (fixed a few merge conflicts) and pushed the new branch here.
When you'll test the NCP, it's important to use the branch of this PR to be sure we are testing against the same code :)

jia200x commented:
@aabadie cool!

Yes, I will continue with this PR branch. Unfortunately I didn't have time to work on it today, and I will be off until Monday

biboc commented:
FYI, OpenThread has ported the board samr21-xpro: 
https://github.com/openthread/openthread/tree/master/examples/platforms/samr21
Anyway, I do think we need OpenThread port on RIOT because OpenThread does not implement any OS functions so RIOT would be good OS to use OpenThread

jia200x commented:
> FYI, OpenThread has ported the board samr21-xpro:
https://github.com/openthread/openthread/tree/master/examples/platforms/samr21
Anyway, I do think we need OpenThread port on RIOT because OpenThread does not implement any OS functions so RIOT would be good OS to use OpenThread

Indeed. I did some experiments with conn_api on top of OpenThread. I ran the posix example with OT as network stack instead of GNRC, and the same application worked for both cases. So, all RIOT apps with 15.4 could run on top of OpenThread :)

aabadie commented:
Hi @jia200x 
> I did some experiments with conn_api on top of OpenThread

Do you mean sock instead of conn_api ? because the latter is not part of RIOT anymore.

> all RIOT apps with 15.4 could run on top of OpenThread :)

That sounds awesome ! Do you plan to provide a PR for that ?
by the way, did you have time for testing this PR ? I wish we could merge it one day :)

jia200x commented:
Hi @aabadie 

> Do you mean sock instead of conn_api ? because the latter is not part of RIOT anymore.

No, [I meant conn_api]. That was looooong time ago (Aug 2016). I currently have some work in sock adoption [here], and I plan to add it after this PR gets merged.

> by the way, did you have time for testing this PR ? I wish we could merge it one day :)

I was able to configure the NCP, call commands with wpantund + wpanctl (scan, etc). I was not able to get the tun interface with wpanctl, but I'm 100% sure is not related to this PR and I will try to make it work when I find a slot (I've been very busy this week because I move in a couple of weeks, but I will try to do it tomorrow :).


[I meant conn_api]:https://github.com/jia200x/RIOT/blob/ac3cff27b780e185e8ca4dd3ffaf1bade2f0fa0e/pkg/openthread/contrib/conn/conn_udp.c
[here]:https://github.com/jia200x/RIOT/tree/openthread_sock_udp/pkg/openthread/contrib/sock

jia200x commented:
There's something wrong with the channel when doing scan. I'm always getting channel 0 (with FTD nodes on channel 26)

`2017-12-15 11:53:33,630 - INFO # | J | Network Name     | Extended PAN     | PAN  | MAC Address      | Ch | dBm | LQI |
2017-12-15 11:53:33,637 - INFO # +---+------------------+------------------+------+------------------+----+-----+-----+
2017-12-15 11:53:33,654 - INFO #  | 0 | codelab          | c0de7ab5c0de7ab5 | c0de | ee58f315a471192f |  0 | -43 |   0 `

I can't join a Thread network following [OpenThread Codelabs], and I think is related to this channel mismatch.

@biboc @aabadie  where you able to join at some point?

[OpenThread Codelabs]:https://codelabs.developers.google.com/codelabs/openthread-hardware/#0

jia200x commented:
Also, ping stops working after some time. I'm experiencing one-way ping between 2 FTDs :(

jia200x commented:
scan breaks ping between nodes. I'm checking what's going on

jia200x commented:
hi @biboc 

Is it OK if I open a new PR based on this one? So I can continue the work there ;)

biboc commented:
@jia200x Yes it's OK

Hyungsin commented:
@jia200x, what is the goal of another PR? 
I think it would be better merge this PR fast even though it is not perfect, as the basic openthread port.
Based on that, several things need to be fixed anyway. I'm keeping those fixes for another PR.

jia200x commented:
Hi @Hyungsin 

I proposed @biboc to open a new PR since he won't have time to merge it. The scope it's just quick fixes and all related to this PR (thread priority, wrong channel during scan), so I guess that will fast up the process of merging. Then, we can add the rest of the features and fixes.

Do you agree?

aabadie commented:
> The scope is just quick fixes and all related to this PR (thread priority, wrong channel during scan)

Maybe you could just add comments in the existing PR or open a PR on the upstream branch ? And we can merge this one when the fixes are applied.

jia200x commented:
> Maybe  you could just add comments in the existing PR or open a PR on the  upstream branch ? And we can merge this one when the fixes are applied.

Ok, let's go that way.

jia200x commented:
So far, I'm able to join OpenThread networks with wpanctl. But ping between laptop and thread network doesn't work. I'm seeing frame rx errors in the NCP.

jia200x commented:
@Hyungsin is ping command working for you through wpantund?

jia200x commented:
I have a branch where I wrote all the fixes I proposed [here].

[here]:https://github.com/jia200x/RIOT/tree/openthread_ftd_ncp

jia200x commented:
It seems the pkg version of OpenThread (and wpantund of last January) don't give IP if the form command is not called from wpanctl. When using form, I can ping to the NCP from the linux computer.
So, I would fix the remaining issues and merge it as is. The latest OpenThread pkg gets IP when calling the join command, [as seen here]

[as seen here]:https://codelabs.developers.google.com/codelabs/openthread-hardware/#7

aabadie commented:
@jia200x I updated the branch with your fix commits. Can you test again (just to be sure) ?

> When using form, I can ping to the NCP from the linux computer.

Is it possible to use this with the actual package version or do we need an update ?

jia200x commented:
@aabadie nice, I will give it a try.

> Is it possible to use this with the actual package version or do we need an update ?

Yes it is, but I still need to figure out how to join the network from the nodes, so that way end devices could be reach from the Linux computer

jia200x commented:
Here is [an updated version of this PR]. After upgrading the pkg, changing NCP port to UART_DEV(1) and fixing some radio issues, ping works between nodes and computer through wpantund.
I tested multihop with IoT-LAB and works as expected.

[an updated version of this PR]:https://github.com/jia200x/RIOT/tree/openthread-FTD-MTD-NCP_V2

aabadie commented:
Weird that Murdock is failing now. Is there something changed in the docker image ? I see this error `automake: not found` in the log. Don't we have this installed ?

aabadie commented:
@jia200x, I integrated your recent changes. I also included a few and removed your UART_DEV update because I think it's useless. Use CFLAGS at build time if you want different values but normally the default should work if the board is plugged by usb and should use STDIO UART.

Could you give another try ?

smlng commented:
Murdock issues are CI related, i.e., missing automake on the workers. We need to update them and add the package, but I wouldn't recommend doing that before the release. So, we need to postpone this into 2018.04

jia200x commented:
There are some issues with the NCP with SAMR21 (EDBG port). When a continuous UART stream exceeds 64 bytes, the NCP node receives a trimmed packet (as seen [here], where Channel0 is the EDBG RX line and Channel1 is TX from the Linux computer)

If I use exactly the same firmware, but connecting an FTDI directly to shared UART pins (PA04 and PA05), it works OK.

If I try to ping with packet size less than 8 bytes, it works through the EDBG port. 
I don't think it's RIOT related, there might be some stty configuration issues.

@Hyungsin what board are you using? Does it work for you?

[here]:https://gist.github.com/jia200x/daac639c022d5b085e0019b63eadbdb0#file-picture-jpg

jia200x commented:
> There are some issues with the NCP with SAMR21 (EDBG port). When a continuous UART stream exceeds 64 bytes, the NCP node receives a trimmed packet (as seen here, where Channel0 is the EDBG RX line and Channel1 is TX from the Linux computer)

Maybe @kaspar030 could give some hints :)

jia200x commented:
>@Hyungsin what board are you using? Does it work for you?

@Hyungsin ping

aabadie commented:
@jia200x, @Hyungsin any updates here ? I think this PR is quite important, it would be great to move forward here. We need to find a solution for the NCP uart issue.

jia200x commented:
@aabadie does it work for you with IoT LAB nodes? This seems to be really Samr21-xpro specific

aabadie commented:
> does it work for you with IoT LAB nodes?

I tested quite some time ago (before the very last changes) and I wasn't able to ping nodes in the network from the Linux A8. With the current status, I wasn't able to configure the NCP, as if wpantund couldn't talk to the NCP.
Maybe you could have a look on your side ?

jia200x commented:
>Maybe you could have a look on your side ?

I will give it a look again. I tend to think it's something with the configuration of wpantund. Because, exactly the same code works with an FTDI.

aabadie commented:
@jia200x, could you find some time to have a look at this one ?

jia200x commented:
@aabadie I'm on it

Hyungsin commented:
@jia200x , @aabadie. 
Sorry I was busy doing other things.

Can I open another PR (a successor of this) which aims to support 
1) SAMR21-XPRO board
2) reflecting the latest openthread code 
3) additional (but necessary) features

Hyungsin commented:
@jia200x, I haven't used the NCP by using EDBG.
But just in case, can you try https://github.com/biboc/RIOT/pull/1?

Also, the latest release was 9 months ago and there have been significant changes since then.
Personally, I think using the master branch (picking up the latest commit) would be better.

biboc commented:
@aabadie @Hyungsin, tell me if you want me to merge https://github.com/biboc/RIOT/pull/1

aabadie commented:
> tell me if you want me to merge

go ahead, that doesn't hurt in the actual state

jia200x commented:
@biboc yes sure, go ahead. Thanks

jia200x commented:
>Also, the latest release was 9 months ago and there have been significant changes since then.
Personally, I think using the master branch (picking up the latest commit) would be better.

Hi @Hyungsin 
Note the OpenThread release is Thread 2017-07-16 compliant, but was released in November-December. There are new features now in this PR, but most of them are already present in the Release.

I will try the NCP with the code you proposed. Thanks!

Hyungsin commented:
@jia200x, @aabadie. Sure I can wait until this PR is merged :)

Hyungsin commented:
@kaspar030, small correction for the title: pkt -> pkg.

kaspar030 commented:
> @kaspar030, small correction for the title: pkt -> pkg.

thx!

jia200x commented:
I'm pretty sure is a Flow Control issue, as seen [here](https://github.com/openthread/openthread/issues/2498). In that case, the fix is not related to this PR.

jia200x commented:
@Hyungsin @aabadie 

Are you OK in splitting this PR in FTD/MTD support and NCP?

- The FTD nodes are working ok and the blocking part is NCP, where there are chances the issue is not RIOT related
- The NCP is just one of the several configurations [as seen in the official openthread site](https://openthread.io/platforms/#host--802154-controller). It's not the only way to provide external IPv6 connectivity (e.g RIOT nodes should be compatible with standard Thread Border Routers, not necessarily related to OpenThread).
- It's possible to use one of the OpenThread platforms as an NCP directly from the OpenThread repository. For instance the samr21-xpro is supported [here](https://github.com/openthread/openthread/tree/master/examples/platforms/samr21). The only benefit of having a RIOT NCP is the fact that it's possible to use other boards. But since there's only support for at86rf2xx now, I don't think it would make a huge difference to postpone a little bit the NCP part.

Also, merging the FTD part can speed up the adoption to other radios, and improve the OpenThread port.

Comments?


- 
-

jia200x commented:
BTW splitting this PR should be fast. All NCP parts are marked via `#ifdef`. 
If you agree with the idea, I would use the OpenThread release (last commit Dec 2017).

Hyungsin commented:
I agree :) Let’s split and merge this!

aabadie commented:
> This is what you mean right? 

@Hyungsin, exactly! The tag name is confusing but the version is from last November. Is this version ok for you ?

Hyungsin commented:
@jia200x, sure.
The version is used in the current code? or do you need to update further?

jia200x commented:
@Hyungsin @aabadie it's quite confusing, since I checked [this](https://github.com/openthread/openthread/commit/a8496c80a117b90a8b3dbbc7233ea52f5841433e) is the last commit of OpenThread (July 2017). However, I still tend to think it's good to keep the Thread Reference implementation (it will be easier for certification, etc).

The current code doesn't run with that version (function prototypes changed) but it's not a big deal to make it work again. There are some features that are not present in the old version (specially CLI) but most important features are there (MTD, SED, Border Router support, etc). @Hyungsin do you have out of your head any features we might miss with the release?

In any case, I'm pretty sure there will come another Release and we will need this PR.

Hyungsin commented:
I think we don't have to worry about features at this time, which can be added in future PRs.
I see this PR as the basic stepping stone, which compiles well when using the 20170716 release.

So if you can work on reprototyping, please go ahead and let it merged :)
The function prototypes have been changed after the release anyway.

After this PR, may be you can keep working on NCP and I can reflect the features I have added so far.

Hyungsin commented:
Any update here?

jia200x commented:
@Hyungsin just opened #9336 . It should be fast to merge since this PR had some reviews.

Hyungsin commented:
@jia200x. Cool!! Thanks!

miri64 commented:
#9336 was merged. So I guess this can be closed.

jia200x commented:
it still has some code to be ported (NCP, etc), but closing it won't hurt anyone :)

miri64 commented:
Then I archive it.
